kong/pdk/vault.lua

```diff
@@ -168,6 +176,71 @@ local function new(self)
   end
 
 
+  local function get_prefix_hash(schema, prefix, base_config)
+    local hash = CONFIG_HASHES[prefix]
+    if hash then
+      return hash
+    end
+
+    for k in schema:each_field() do
+      local v = base_config[k]
+      if v ~= nil then
+        if not hash then
+          hash = true
+          KEY_BUFFER:reset()
+        end
+        KEY_BUFFER:putf("%s=%s;", k, v)
+      end
+    end
+
+    if hash then
+      hash = encode_base64url(md5_bin(KEY_BUFFER:get()))
+      CONFIG_HASHES[prefix] = hash
+    end
+
+    return hash
+  end
+
+
+  local function get_config(schema, base_config, config_overrides, prefix)
+    if not config_overrides or isempty(config_overrides) then
+      if not prefix then
+        return base_config
+      end
+
+      return base_config, get_prefix_hash(schema, prefix, base_config)
+    end
+
+    local config
+    for k, f in schema:each_field() do
+      local v = config_overrides[k]
+      v = arguments.infer_value(v, f)
+      -- TODO: should we be more visible with validation errors?
+      if v ~= nil and schema:validate_field(f, v) then
+        if not config then
+          config = clone(base_config)
+          KEY_BUFFER:reset()
+          if prefix then
+            local hash = get_prefix_hash(schema, prefix, config)
+            if hash then
+              KEY_BUFFER:putf("%s;", hash)
+            end
+          end
```

Comment:
There's a lot of very dense business logic and data manipulation happening in `get_prefix_hash`, `get_config`, and `process_secret`. The code is well-written, so the procedural bits are clear and relatively easy to follow, but the underlying business logic is not always fully clear to me.

This specific branch I highlighted is a good example. To my eyes it looks like a potential bug, and I immediately have the following questions:

1. Is it correct to be taking a hash of `config`, or should we be hashing `base_config` instead?
2. Is it okay that we're doing this in the middle of iteration? Couldn't it lead to ordering problems where we have two hashes that are equivalent but don't compare equally (`foo=bar;<my-hash>;baz=bat` vs `<my-hash>;foo=bar;baz=bat`)

Maybe these are not worthwhile concerns, but as the reader I cannot say for sure. I think this is a good case for adding some extra doc strings and explanatory comments so that the business logic is a little bit more obvious.

Reply:
The base config of process secrets never changes after kong is started. So for that we don't need to hash base config. The config of vault entity can change as admin api is used. The prefix is only available with `config` secrets, aka when using vault entities.